File: mne_brain/src/mne_brain/ica/ica_classification.py
"""ICA component feature extraction and artifact suggestions."""

from __future__ import annotations

import numpy as np
import pandas as pd


def compute_ica_features(
    component_maps: np.ndarray,
    sources: np.ndarray,
    fs: float,
    channel_names: list[str],
) -> pd.DataFrame:
    """Compute NeuroMIND-style ICA features for each component."""
    n_channels, n_components = component_maps.shape
    frontal_idx = [
        idx
        for idx, ch in enumerate(channel_names)
        if ch.upper().startswith("FP") or ch.upper().startswith("F")
    ]
    temporal_idx = [
        idx
        for idx, ch in enumerate(channel_names)
        if ch.upper().startswith("T") or ch.upper().startswith("TP")
    ]
    if not frontal_idx:
        frontal_idx = list(range(min(4, n_channels)))
    if not temporal_idx:
        temporal_idx = list(range(max(0, n_channels - 4), n_channels))

    all_idx = np.arange(n_channels)
    nonfrontal_idx = np.setdiff1d(all_idx, frontal_idx)
    nontemporal_idx = np.setdiff1d(all_idx, temporal_idx)

    rows = []
    for comp in range(n_components):
        spatial = component_maps[:, comp]
        source = sources[comp]

        frontal_ratio = _safe_mean_abs(spatial[frontal_idx]) / (
            _safe_mean_abs(spatial[nonfrontal_idx]) + 1e-12
        )
        temporal_ratio = _safe_mean_abs(spatial[temporal_idx]) / (
            _safe_mean_abs(spatial[nontemporal_idx]) + 1e-12
        )
        blink_ratio = _bandpower(source, fs, 0.5, 4.0) / (_bandpower(source, fs, 4.0, 40.0) + 1e-12)
        emg_ratio = _bandpower(source, fs, 30.0, 80.0) / (_bandpower(source, fs, 1.0, 30.0) + 1e-12)
        line_ratio = _bandpower(source, fs, 48.0, 52.0) / (_bandpower(source, fs, 0.5, 100.0) + 1e-12)
        kurt = _kurtosis(source)
        std = source.std(ddof=1)
        extreme_frac = 0.0 if std == 0 else float(np.mean(np.abs(source - source.mean()) > 5 * std))

        rows.append(
            {
                "component": comp,
                "frontal_ratio": round(float(frontal_ratio), 4),
                "temporal_ratio": round(float(temporal_ratio), 4),
                "blink_ratio": round(float(blink_ratio), 4),
                "emg_ratio": round(float(emg_ratio), 4),
                "line_ratio": round(float(line_ratio), 6),
                "kurtosis": round(float(kurt), 4),
                "extreme_frac": round(float(extreme_frac), 6),
            }
        )
    return pd.DataFrame(rows)


def evaluate_ica_components(features: pd.DataFrame, *, artifact_thresh: float = 1.5) -> pd.DataFrame:
    """Add artifact scores and labels to an ICA feature table."""
    result = features.copy()
    fz = _zscore(result["frontal_ratio"].to_numpy())
    tz = _zscore(result["temporal_ratio"].to_numpy())
    blz = _zscore(result["blink_ratio"].to_numpy())
    emz = _zscore(result["emg_ratio"].to_numpy())
    lnz = _zscore(result["line_ratio"].to_numpy())
    kz = _zscore(result["kurtosis"].to_numpy())
    exz = _zscore(result["extreme_frac"].to_numpy())

    ocular = 0.4 * fz + 0.4 * blz + 0.2 * kz
    muscle = 0.5 * emz + 0.3 * tz + 0.2 * kz
    line = lnz
    jump = 0.5 * kz + 0.5 * exz

    scores = np.vstack([ocular, muscle, line, jump]).T
    labels = []
    global_score = []
    label_names = ["eye/blink", "muscle", "line_noise", "jump"]
    for row in scores:
        max_idx = int(np.argmax(row))
        score = float(row[max_idx])
        global_score.append(score)
        labels.append(label_names[max_idx] if score > artifact_thresh else "brain")

    result["ocular_score"] = np.round(ocular, 3)
    result["muscle_score"] = np.round(muscle, 3)
    result["line_score"] = np.round(line, 3)
    result["jump_score"] = np.round(jump, 3)
    result["artifact_score"] = np.round(global_score, 3)
    result["artifact_type"] = labels
    return result


def suggest_rejected_components(evaluated: pd.DataFrame) -> list[int]:
    """Return non-brain component indices suggested for rejection."""
    bad = evaluated.loc[evaluated["artifact_type"] != "brain", "component"]
    return [int(idx) for idx in bad.to_list()]


def mne_artifact_suggestions(ica, raw) -> dict[str, list[int]]:
    """Try MNE's built-in EOG and muscle detectors when possible."""
    suggestions: dict[str, list[int]] = {"eog": [], "muscle": []}
    try:
        if "Fp2" in raw.ch_names:
            eog_idx, _ = ica.find_bads_eog(raw, ch_name="Fp2", verbose="ERROR")
            suggestions["eog"] = [int(idx) for idx in eog_idx]
    except Exception:
        suggestions["eog"] = []

    try:
        muscle_idx, _ = ica.find_bads_muscle(raw, verbose="ERROR")
        suggestions["muscle"] = [int(idx) for idx in muscle_idx]
    except Exception:
        suggestions["muscle"] = []
    return suggestions


def _safe_mean_abs(values: np.ndarray) -> float:
    return float(np.mean(np.abs(values))) if values.size else 0.0


def _zscore(values: np.ndarray) -> np.ndarray:
    std = values.std(ddof=1)
    if std == 0 or np.isnan(std):
        return np.zeros_like(values, dtype=np.float64)
    return (values - values.mean()) / std


def _kurtosis(values: np.ndarray) -> float:
    centered = values - values.mean()
    variance = np.mean(centered**2)
    if variance == 0:
        return 0.0
    return float(np.mean(centered**4) / (variance**2))


def _bandpower(values: np.ndarray, fs: float, f_lo: float, f_hi: float) -> float:
    x = values - values.mean()
    freqs = np.fft.rfftfreq(x.size, d=1.0 / fs)
    psd = np.abs(np.fft.rfft(x)) ** 2 / (fs * x.size)
    idx = (freqs >= f_lo) & (freqs <= f_hi)
    return float(psd[idx].sum())

File: src/ica/ICAClassification.jl
# NeuroMIND/src/ica/ICAClassification.jl
#
# Clasificación automática de componentes ICA.
# Portado de EEG_Julia/src/ICA/ICA_cleaning.jl (sin Plots, Julia puro).
#
# Features por componente:
#   frontal_ratio, temporal_ratio, blink_ratio, emg_ratio, line_ratio,
#   kurtosis, extreme_frac
# Scores:
#   ocular_score, muscle_score, line_score, jump_score → artifact_score
# Etiquetas: "brain", "eye/blink", "muscle", "line_noise", "jump", "unknown"

# ─── Auxiliares ────────────────────────────────────────────────

function _zscore(v::AbstractVector)
    m = mean(v); s = std(v)
    s == 0.0 && return zeros(length(v))
    return (v .- m) ./ s
end

function _kurtosis_simple(x::AbstractVector)
    m  = mean(x); xc = x .- m
    s2 = mean(xc .^ 2)
    s2 == 0.0 && return 0.0
    return mean(xc .^ 4) / (s2^2)
end

function _bandpower(x::AbstractVector, fs::Real, f_lo::Real, f_hi::Real)
    x  = x .- mean(x)
    N  = length(x)
    X  = rfft(x)
    freqs = (0:length(X)-1) .* (fs / N)
    psd = abs.(X) .^ 2 ./ (fs * N)
    return sum(psd[(freqs .>= f_lo) .& (freqs .<= f_hi)])
end

# ─── Features por IC ───────────────────────────────────────────

"""
    compute_ica_features(A, S, fs, ch_names) -> DataFrame

Calcula 7 features por componente ICA.
  - frontal_ratio, temporal_ratio: distribución espacial (A)
  - blink_ratio, emg_ratio, line_ratio: potencias de banda (S)
  - kurtosis, extreme_frac: estadísticos temporales (S)
"""
function compute_ica_features(
    A::AbstractMatrix{Float64},
    S::AbstractMatrix{Float64},
    fs::Real,
    ch_names::Vector{String}
)::DataFrame

    n_ch, n_ic = size(A)
    all_idx = 1:n_ch

    # Índices de canales frontales y temporales (nomenclatura 10-20)
    frontal_idx  = [i for (i, ch) in enumerate(ch_names)
                    if occursin(r"^F[PC0-9ZCT]?"i, uppercase(ch)) ||
                       startswith(uppercase(ch), "FP")]
    temporal_idx = [i for (i, ch) in enumerate(ch_names)
                    if startswith(uppercase(ch), "T") ||
                       startswith(uppercase(ch), "TP")]

    if isempty(frontal_idx);  frontal_idx  = collect(1:min(4, n_ch)); end
    if isempty(temporal_idx); temporal_idx = collect(max(1, n_ch-3):n_ch); end

    nonfrontal_idx  = setdiff(all_idx, frontal_idx)
    nontemporal_idx = setdiff(all_idx, temporal_idx)

    fr   = zeros(n_ic); tr   = zeros(n_ic)
    blnk = zeros(n_ic); emg  = zeros(n_ic)
    lnr  = zeros(n_ic); kurt = zeros(n_ic)
    exfr = zeros(n_ic)

    n_samp = size(S, 2)

    for k in 1:n_ic
        map_k = A[:, k]
        s_k   = S[k, :]

        # Spatial ratios
        fr[k] = mean(abs.(map_k[frontal_idx]))  /
                (mean(abs.(map_k[nonfrontal_idx]))  + 1e-12)
        tr[k] = mean(abs.(map_k[temporal_idx])) /
                (mean(abs.(map_k[nontemporal_idx])) + 1e-12)

        # Spectral ratios
        P04   = _bandpower(s_k, fs, 0.5,  4.0)
        P440  = _bandpower(s_k, fs, 4.0, 40.0)
        P130  = _bandpower(s_k, fs, 1.0, 30.0)
        P3080 = _bandpower(s_k, fs, 30.0, 80.0)
        Ptot  = _bandpower(s_k, fs, 0.5, 100.0)
        P4852 = _bandpower(s_k, fs, 48.0, 52.0)

        blnk[k] = P04   / (P440  + 1e-12)
        emg[k]  = P3080 / (P130  + 1e-12)
        lnr[k]  = P4852 / (Ptot  + 1e-12)

        # Statistical
        kurt[k] = _kurtosis_simple(s_k)
        μ = mean(s_k); σ = std(s_k)
        exfr[k] = σ == 0.0 ? 0.0 :
            count(t -> abs(t - μ) > 5σ, s_k) / n_samp
    end

    return DataFrame(
        component       = 1:n_ic,
        frontal_ratio   = round.(fr,   digits=4),
        temporal_ratio  = round.(tr,   digits=4),
        blink_ratio     = round.(blnk, digits=4),
        emg_ratio       = round.(emg,  digits=4),
        line_ratio      = round.(lnr,  digits=6),
        kurtosis        = round.(kurt, digits=4),
        extreme_frac    = round.(exfr, digits=6),
    )
end

# ─── Scores y clasificación ────────────────────────────────────

"""
    evaluate_ica_components(feat_df; artifact_thresh=1.5) -> DataFrame

Añade scores y etiqueta al DataFrame de features.
Etiquetas: "brain", "eye/blink", "muscle", "line_noise", "jump", "unknown"
"""
function evaluate_ica_components(
    feat_df::DataFrame;
    artifact_thresh::Real = 1.5
)::DataFrame

    fz  = _zscore(feat_df.frontal_ratio)
    tz  = _zscore(feat_df.temporal_ratio)
    blz = _zscore(feat_df.blink_ratio)
    emz = _zscore(feat_df.emg_ratio)
    lnz = _zscore(feat_df.line_ratio)
    kz  = _zscore(feat_df.kurtosis)
    exz = _zscore(feat_df.extreme_frac)

    n_ic = nrow(feat_df)

    ocular  = @. 0.4*fz + 0.4*blz + 0.2*kz
    muscle  = @. 0.5*emz + 0.3*tz + 0.2*kz
    line    = lnz
    jump    = @. 0.5*kz + 0.5*exz

    artifact_global = zeros(n_ic)
    labels          = Vector{String}(undef, n_ic)

    for k in 1:n_ic
        scores_k = (ocular[k], muscle[k], line[k], jump[k])
        mx = maximum(scores_k)
        artifact_global[k] = mx
        if mx > artifact_thresh
            labels[k] = if mx == ocular[k]; "eye/blink"
                        elseif mx == muscle[k]; "muscle"
                        elseif mx == line[k]; "line_noise"
                        else "jump"
                        end
        else
            labels[k] = "brain"
        end
    end

    result = copy(feat_df)
    result[!, :ocular_score]   = round.(ocular, digits=3)
    result[!, :muscle_score]   = round.(muscle, digits=3)
    result[!, :line_score]     = round.(line,   digits=3)
    result[!, :jump_score]     = round.(jump,   digits=3)
    result[!, :artifact_score] = round.(artifact_global, digits=3)
    result[!, :artifact_type]  = labels
    return result
end
